server/channels/db/migrations/mysql/000111_create_retentionidsfordeletion_table.up.sql

```diff
@@ -0,0 +1,7 @@
+CREATE TABLE IF NOT EXISTS RetentionIdsForDeletion (
+    Id varchar(26) NOT NULL,
+    TableName varchar(64),
+    Ids json,
```

Comment:
Would this be better as a record per table/id versus an unbounded number of Ids stuffed into a json field?

Reply:
We did discuss this in the spec I previously worked on https://docs.google.com/document/d/17IIomw1cf1P0hvYEdmtvk4cvy-m0L_G6RpC4QemDGUc/edit?disco=AAAA09pw-F4. The final code changed quite a bit compared to the spec but the table structure stayed the same.

Echoing Agniva's comment: The drawbacks of doing one row per table/id is the writes will be much slower as in a single query, you might delete 100 rows, but then you would need to insert 100 rows separately, or create a batch insert query to insert 100 rows at once. Also, cleaning up would be slower as well. The advantage though is that it's more reliable to manage the batches compared to the unbounded arrays.

The concerns about using a JSON or array field is that as you mentioned it's unbounded right now, we could set a max batch size for the data retention batch size config? Also, one thing that could become an issue:
1. You start the data retention job with a batch size of 3000
2. You have inserted a few rows into RetentionIdsForDeletion
3. The job fails due to some db timeouts before using the inserted rows
4. You decide a batch number of 3000 is too high so you reduce it to 1000
5. The issue now is that you have some batches of 3000 lingering in this table that will be used to delete reactions.

The reactions delete query doesn't have any joins and is a lot more efficient than others but the above scenario is a slight concern of mine.

I'm open to discussion if we want to change it, it wouldn't be a major code refactor

Reply:
I've not done any performance testing around these myself, but I do know that for MySQL the json is stored like blob.  When inserting new ids into the JSON it requires a full re-write of the blob.  Certainly it would be more comfortable if that was a bounded list of ids so we can predict the impact of very large retention jobs.  The 5 step issue above is because there is no easy way to count the items in the blob. IMO that's just another reason not to use json here, but if this has been discussed at length and agreed to, I won't block it.  

One general point:  it is not nearly as important how long the retention job takes than keeping the impact on users to a minimum.  Do the work in small batches that take longer overall but don't stress the DB or app servers.  I notice a lot of these apis are fetching potentially large sets of IDs then spinning hard in a loop  on the app server.  These are pre-existing to your PR.  That pattern will not help us scale.

Reply:
Sorry I think I misunderstood a little, we never actually edit any of the json blobs/arrays directly. All we do is insert the array and it's size will match whatever you have set for DataRetention.BatchSize. Then when we are deleting records using these arrays it's an all or nothing transaction. Eg: `Delete From Reactions where postId In (<arrayOfIds>)`.

So we never have to search/insert/delete from the JSON blob